### ISSUE TITLE
job submissions to wrong partition on Hortense

### ISSUE THREAD

**Issue by laraPPr** [EESSI/test-suite#68]
Reframe submits all the tests to the same partition. So if reframe is started from the `cpu_milan` partitions all the test that are found for other partitions will also be submitted to `cpu_milan`. This especially goes horribly wrong when starting from a GPU partition. Since all the tests meant for the cpu-partitions fail mediately.

We have narrowed down the problem to the following parts of the hortense system and the `vsc_hortense.py` config file:

- The config file adds the following line, `#SBATCH --partition=cpu_milan` in the job script ( `rfm_job.sh`)
- We have the following environment variable, `SBATCH_PARTITION=cpu_rome`, set by Hortense cluster module
- Reframe submits the jobs with `sbatch rfm_job.sh`
- So the `SBATCH_PARTITION` variable wins

Could it be possible that reframe submits the job with `sbatch --partition=cpu_milan rfm.job`?

A possible work around might be to use `prepare_cmds` to set the environment variable `SBATCH_PARTITION` for every partition in `config/vsc_hortense.py`.

**Comment by boegel** [comment]
@vkarak Thoughts on this? Is it intentional that having `$SBATCH_PARTITION` set overrules the ReFrame configuration?

**Comment by vkarak** [comment]
No, it's not intentional, we haven't thought about this type of set up, actually :-) But since this is really related to how the target system is set up, I think that setting the `prepare_cmds` in the configuration of this partition as @laraPPr suggests is the right way. I have an additional question: where is the Hortense cluster module loaded? Is it also listed in the partition config? If that's the case, then perhaps omitting the `--partition` from the partition's `sched_access` is probably another solution, as setting the partition is taken care of by the module already.

**Comment by casparvl** [comment]
@vkarak Aren't the `prepare_cmds` only executed on the batch node? They aren't executed on the submit host (i.e. where one is running the `reframe` command), right?

We just tried something else together with Lara: we unloaded one of their sticky environment modules (the one that sets the `SBATCH_PARTITION` variable), then run the `reframe`. That works, since there is no conflict anymore between the environment variable and the `#SBATCH --partition` line that ReFrame puts in the batch script.

The only thing is, I don't know what unintended side effects the unload of that sticky module might have, but @boegel probably knows... :)

**Comment by boegel** [comment]
@vkarak Not being able to add options directly to the `sbatch` command is actually not a problem, as long as we can modify the environment in which the `sbatch` command is run (so we can make sure that `$SBATCH_PARTITION` is either unset, or that it's set correctly).

Is there a way to do that?

Unloading the module (with force) that sets `$SBATCH_PARTITION` works, but can we specify that in the config file somehow, or does that need to be done before running `reframe`?

(it's actually the same question twice, sort of...)

**Comment by vkarak** [comment]
I had a look again into it and indeed you can't cleanly modify the environment where `sbatch` will run, especially per partition. You can only modify globally the environment that reframe will run in. I suggest opening an issue for this, as it requires some thinking on how to do this properly.

**Comment by boegel** [comment]
https://github.com/reframe-hpc/reframe/issues/2970 was closed just now, so what we need is coming in a ReFrame release soon?

**Comment by vkarak** [comment]
Yes, it is already merged in `develop` if you would like to try it out and is scheduled for 4.7.

**Comment by vkarak** [comment]
@laraPPr Was this regression in 4.7.2 only? Does it work in 4.7.0?

**Comment by vkarak** [comment]
Could you also try 4.7.1? Because the only chance to have broken is by the [fixes](https://github.com/reframe-hpc/reframe/pull/3352) introduced in 4.7.2. Also, could you describe the exact scenario you are trying and what would be the expected behaviour?

**Comment by smoors** [comment]
> ```
> 'sched_options': {
>    'shed_access_in_submit': True,
> },
> ```

as Caspar found out, you have a typo, should be `sched_access_in_submit`

**Comment by laraPPr** [comment]
The problem was indeed the typo tested with 4.7.0, 4.7.1 and 4.72.

**Comment by laraPPr** [comment]
https://github.com/EESSI/test-suite/pull/225

**Comment by laraPPr** [comment]
@Casparvl can this one be closed (Since it is resolved when using ReFrame 4.7.0) or should we wait until the CI is updated to use ReFrame 4.7.0 or older?
